platforms/jvm/plugins-java-base/src/integTest/groovy/org/gradle/api/plugins/jvm/internal/DefaultJvmFeatureTest.groovy

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.plugins.jvm.internal
+
+import org.gradle.api.InvalidUserDataException
+import org.gradle.api.plugins.JavaBasePlugin
+import org.gradle.api.plugins.JavaPluginExtension
+import org.gradle.test.fixtures.AbstractProjectBuilderSpec
+
+class DefaultJvmFeatureTest extends AbstractProjectBuilderSpec {
+    // TODO: This test definitely isn't appropriate for Component any more, so I've moved it here.
+    // Should features create the sourcesets they are going to use?  If they are being passed in
```

Comment:
I think eventually we want to move away from source sets entirely. They are a weird mixture of features and targets and they do not really fit into our new model. 

For example, the source set effectively owns the `compileJava` and `jar` tasks that make up a compilation/target. So, really, it is the target that should own the source set. Eventually, the source set should disappear and all we have is the target. 

For that reason, I think it is best to encapsulate the existence of the source set, so that we don't rely on it existing. It is only there for compatibility since that is all the public API has access to currently.

Where this gets difficult is naming, which is why the existing code had the name provided but delegated the construction of the source set to the component. 

Ideally, the feature would have instantiated the source set itself, but due to the way `registerFeature` works, the source set needs to be provided externally.

Reply:
> I think eventually we want to move away from source sets entirely.

I think the _current_ implementation of source sets we have absolutely tries to do too much when many times only a small fraction of its functionality is needed.  I think a minimal source set (that merely defines a collection of directories in a particular source language, without concerning itself with how to compile or run that source) is a necessary software definition "primitive".

> For that reason, I think it is best to encapsulate the existence of the source set, so that we don't rely on it existing.

I'm open to trying this, but it's out of scope for this change.  I thought about having the feature optionally create a source set if none was passed, but that seems like too big a change in this simple internal refactoring.  For now, we make a source set, which is used to make a feature, which is used to make a component.

> due to the way registerFeature works, the source set needs to be provided externally.

Yep.  Let's worry about this later.  This is just a small internal improvement for now.